### PR TITLE
API changes and additions

### PR DESCRIPTION
**Pull request by saynb**
* Correcting match_type enum names to match convention to tdi_tna_match_type_e
* Adding tdi_tna_target_e
* Adding const to Target::getValue
* Correcting tdi_flags_e enum to correctly represent the 64 bit value
* Adding p4_name to TdiInfo class, description and API to retrieve
* Adding backpointer to TdiInfo in Table class
* converting TableData:actionIdGet() to return const ref to actionId itself rather than return status and take in param
* making table_, learn_, and all_fields_set_ in TableData to be protected so that derived classes can have access and change them freely. Future change = add correct reset and default initialize functions so that only one entry point to change these internal member variables during init time or reset time
* Adding match_type information in KeyFieldValue so that TableKey setValue can know the object type during runtime without dynamic_cast
* Converting tdi::Flags to a purely public class. Any device can directly call Flags(uint64_t) and create a Flag object to keep it simple. Device::createFlags() won't be needed and will be removed.
* Adding temp actionIdListGet() function. Future imrovement would be to remove any kind of vector copying and just store required vectors in memory and return const refs.

### CODE REVIEW

File: include/tdi/arch/tna/tna_defs.h [hunk 1]
```diff
@@ -50,9 +50,14 @@ typedef enum {
  * @brief Key Field Match Type. A key can have multiple fields,
  * each with a different match type
  */
-enum tdi_match_type_tna_e {
-  TDI_MATCH_TYPE_TNA_RANGE = TDI_MATCH_TYPE_ARCH,
-  TDI_MATCH_TYPE_TNA_ATCAM,
+enum tdi_tna_match_type_e {
+  TDI_TNA_MATCH_TYPE_RANGE = TDI_MATCH_TYPE_ARCH,
```
Comment (weiqiand) [inline include/tdi/arch/tna/tna_defs.h:54]: The tdi_match_type_core_e as we discussed early, have the TDI_MATCH_TYPE_RANGE defined. Not sure TNA can use it or not. Here is the latest list.
 41 enum tdi_match_type_core_e {
 42   TDI_MATCH_TYPE_EXACT = TDI_MATCH_TYPE_CORE,
 43   TDI_MATCH_TYPE_TERNARY,
 44   TDI_MATCH_TYPE_RANGE,
 45   TDI_MATCH_TYPE_LPM,
 46 };

File: src/tdi_table_data.cpp [hunk 2]
```diff
@@ -134,11 +134,6 @@ tdi_status_t TableData::getValue(const tdi_id_t & /*field_id*/,
   return TDI_NOT_SUPPORTED;
 }
 
-tdi_status_t TableData::actionIdGet(tdi_id_t *act_id) const {
```
Comment (weiqiand) [inline src/tdi_table_data.cpp:137]: Why we delete actionIdGet from the TableData?

File: include/tdi/arch/tna/tna_defs.h (hunk 1 above)
Comment (saynb) [reply include/tdi/arch/tna/tna_defs.h:54]: That looks good. You can integrate that as part of your PR itself

File: src/tdi_table_data.cpp (hunk 2 above)
Comment (saynb) [reply src/tdi_table_data.cpp:137]: Made it simpler. Earlier we were returning TDI_SUCCESS every time. Now we just return the action ID directly

**Review by weiqiand — APPROVED** [review]

File: include/tdi/arch/tna/tna_defs.h (hunk 1 above)
Comment (weiqiand) [reply include/tdi/arch/tna/tna_defs.h:54]: ok thanks

File: src/tdi_table_data.cpp (hunk 2 above)
Comment (weiqiand) [reply src/tdi_table_data.cpp:137]: ok. Then I need to change my calling function ...

File: include/tdi/common/tdi_table.hpp [hunk 3]
```diff
@@ -650,10 +650,16 @@ class Table {
    */
   virtual bool actionIdApplicable() const { return false; }
 
+  const TdiInfo *tdiInfoGet() const { return tdi_info_; };
+
  protected:
-  Table(const TableInfo *table_info) : table_info_(table_info){};
+  Table(const TdiInfo *tdi_info, const TableInfo *table_info)
```
[the remark this answers is not in the document]
Comment (saynb) [reply include/tdi/common/tdi_table.hpp:656]: yes

**Pull request merged**